common/src/main/java/io/druid/math/expr/Function.java

```diff
@@ -615,18 +684,122 @@ public String name()
     }
 
     @Override
-    public Number apply(List<Expr> args, Expr.ObjectBinding bindings)
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
     {
       if (args.size() != 3) {
-        throw new RuntimeException("function 'if' needs 3 argument");
+        throw new IllegalArgumentException("function 'if' needs 3 argument");
       }
 
-      Number x = args.get(0).eval(bindings);
-      if (x instanceof Long) {
-        return x.longValue() > 0 ? args.get(1).eval(bindings) : args.get(2).eval(bindings);
-      } else {
-        return x.doubleValue() > 0 ? args.get(1).eval(bindings) : args.get(2).eval(bindings);
+      ExprEval x = args.get(0).eval(bindings);
+      return x.asBoolean() ? args.get(1).eval(bindings) : args.get(2).eval(bindings);
+    }
+  }
+
+  class CastFunc extends DoubleParam
+  {
+    @Override
+    public String name()
+    {
+      return "cast";
+    }
+
+    @Override
+    protected ExprEval eval(ExprEval x, ExprEval y)
+    {
+      String castTo = y.stringValue();
+      if ("string".equals(castTo)) {
+        return ExprEval.of(x.value() == null ? null : String.valueOf(x.value()));
+      }
+      if ("long".equals(castTo)) {
+        return ExprEval.of(x.type() == ExprType.STRING ? Long.valueOf(x.stringValue()) : x.longValue());
+      }
+      if ("double".equals(castTo)) {
+        return ExprEval.of(x.type() == ExprType.STRING ? Double.valueOf(x.stringValue()) : x.doubleValue());
+      }
+      throw new IllegalArgumentException("invalid type " + castTo);
+    }
+  }
+
+  class TimestampFromEpochFunc implements Function
```

Comment:
Druid favors JodaTime over java.util.Date. Any reason we can't implement this method using JodaTime?